credentials/credentials_test.go

```diff
@@ -59,3 +63,93 @@ func TestTLSClone(t *testing.T) {
 		t.Fatalf("Change in clone should not affect the original, c.Info().ServerName = %v, want %v", c.Info().ServerName, expectedServerName)
 	}
 }
+
+func TestTLSClientHandshakeReturnsAuthInfo(t *testing.T) {
+	localPort := ":5050"
+	tlsDir := "../test/testdata/"
+	lis, err := net.Listen("tcp", localPort)
+	if err != nil {
+		t.Fatalf("Failed to start local server. Listener error: %v", err)
+	}
+	serverTLS, err := NewServerTLSFromFile(tlsDir+"server1.pem", tlsDir+"server1.key")
+	if err != nil {
+		t.Fatalf("Failed to create server TLS. Error: %v", err)
+	}
+	var serverAuthInfo AuthInfo
+	done := make(chan bool)
+	go func() {
+		defer func() {
+			done <- true
+		}()
+		serverRawConn, _ := lis.Accept()
+		serverConn := tls.Server(serverRawConn, serverTLS.(*tlsCreds).config)
+		serverErr := serverConn.Handshake()
+		if serverErr != nil {
+			t.Fatalf("Error on server while handshake. Error: %v", serverErr)
+		}
+		serverAuthInfo = TLSInfo{serverConn.ConnectionState()}
+	}()
+	defer lis.Close()
```

Comment:
Move this to the place right after listen().
If something fails before this line, lis will not be closed().